File: src/test/java/org/telran/lecture_13_avl/practice/AvlTreeTest.java
package org.telran.lecture_13_avl.practice;


import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.NoSuchElementException;

import static org.junit.jupiter.api.Assertions.*;

class AvlTreeTest {
    private AvlTree tree;

    @BeforeEach
    void setUp() {
        tree = new AvlTree();
    }

    // Вспомогательные методы для проверки баланса дерева
    private int height(AVLNode node) {
        return node == null ? 0 : node.getHeight();
    }

    private int balanceFactor(AVLNode node) {
        return node == null ? 0 : height(node.getLeft()) - height(node.getRight());
    }

    @Test
    void testEmptyTree() {
        assertTrue(tree.isEmpty());
        assertEquals(0, tree.size());
        assertThrows(NoSuchElementException.class, () -> tree.min());
        assertThrows(NoSuchElementException.class, () -> tree.max());
        assertFalse(tree.contains(5));
        assertNull(tree.search(5));
    }

    @Test
    void testInsertSingleNode() {
        tree.insert(5);

        assertFalse(tree.isEmpty());
        assertEquals(1, tree.size());
        assertEquals(5, tree.min());
        assertEquals(5, tree.max());
        assertTrue(tree.contains(5));
        assertNotNull(tree.search(5));
        assertEquals(5, tree.search(5).getValue());
        assertTrue(tree.isBalanced(tree.getRoot()));
    }

    @Test
    void testInsertMultipleNodes() {
        tree.insert(5);
        tree.insert(3);
        tree.insert(7);
        tree.insert(2);
        tree.insert(4);
        tree.insert(6);
        tree.insert(8);

        assertEquals(7, tree.size());
        assertEquals(2, tree.min());
        assertEquals(8, tree.max());
        assertTrue(tree.contains(3));
        assertTrue(tree.contains(7));
        assertFalse(tree.contains(1));
        assertFalse(tree.contains(9));
        assertTrue(tree.isBalanced(tree.getRoot()));
    }

    @Test
    void testInsertDuplicates() {
        tree.insert(5);
        tree.insert(5);
        tree.insert(5);

        assertEquals(1, tree.size());
        assertEquals(5, tree.min());
        assertEquals(5, tree.max());
    }

    @Test
    void testLeftRotation() {
        // Right-heavy scenario that should trigger left rotation
        tree.insert(10);
        tree.insert(20);
        tree.insert(30);

        // Check if the tree is balanced
        AVLNode root = tree.getRoot();
        assertEquals(20, root.getValue());
        assertEquals(10, root.getLeft().getValue());
        assertEquals(30, root.getRight().getValue());
        assertTrue(tree.isBalanced(root));
    }

    @Test
    void testRightRotation() {
        // Left-heavy scenario that should trigger right rotation
        tree.insert(30);
        tree.insert(20);
        tree.insert(10);

        // Check if the tree is balanced
        AVLNode root = tree.getRoot();
        assertEquals(20, root.getValue());
        assertEquals(10, root.getLeft().getValue());
        assertEquals(30, root.getRight().getValue());
        assertTrue(tree.isBalanced(root));
    }

    @Test
    void testLeftRightRotation() {
        // Scenario that should trigger a left-right rotation
        tree.insert(30);
        tree.insert(10);
        tree.insert(20);

        // Check if the tree is balanced
        AVLNode root = tree.getRoot();
        assertEquals(20, root.getValue());
        assertEquals(10, root.getLeft().getValue());
        assertEquals(30, root.getRight().getValue());
        assertTrue(tree.isBalanced(root));
    }

    @Test
    void testRightLeftRotation() {
        // Scenario that should trigger a right-left rotation
        tree.insert(10);
        tree.insert(30);
        tree.insert(20);

        // Check if the tree is balanced
        AVLNode root = tree.getRoot();
        assertEquals(20, root.getValue());
        assertEquals(10, root.getLeft().getValue());
        assertEquals(30, root.getRight().getValue());
        assertTrue(tree.isBalanced(root));
    }

    @Test
    void testRemoveSingleNode() {
        tree.insert(5);
        tree.remove(5);

        assertTrue(tree.isEmpty());
        assertEquals(0, tree.size());
        assertFalse(tree.contains(5));
    }

    @Test
    void testRemoveNonExistentNode() {
        tree.insert(5);
        tree.remove(10);

        assertEquals(1, tree.size());
        assertTrue(tree.contains(5));
    }

    @Test
    void testRemoveLeafNode() {
        tree.insert(20);
        tree.insert(10);
        tree.insert(30);
        tree.insert(5);

        assertEquals(4, tree.size());
        tree.remove(5);

        assertEquals(3, tree.size());
        assertFalse(tree.contains(5));
        assertTrue(tree.isBalanced(tree.getRoot()));
    }

    @Test
    void testRemoveNodeWithOneChild() {
        tree.insert(20);
        tree.insert(10);
        tree.insert(30);
        tree.insert(5);

        assertEquals(4, tree.size());
        tree.remove(10);

        assertEquals(3, tree.size());
        assertFalse(tree.contains(10));
        assertTrue(tree.contains(5));
        assertTrue(tree.isBalanced(tree.getRoot()));
    }

    @Test
    void testRemoveNodeWithTwoChildren() {
        tree.insert(20);
        tree.insert(10);
        tree.insert(30);
        tree.insert(5);
        tree.insert(15);

        assertEquals(5, tree.size());
        tree.remove(10);

        assertEquals(4, tree.size());
        assertFalse(tree.contains(10));
        assertTrue(tree.contains(5));
        assertTrue(tree.contains(15));
        assertTrue(tree.isBalanced(tree.getRoot()));
    }

    @Test
    void testRemoveRoot() {
        tree.insert(20);
        tree.insert(10);
        tree.insert(30);

        assertEquals(3, tree.size());
        tree.remove(20);

        assertEquals(2, tree.size());
        assertFalse(tree.contains(20));
        assertTrue(tree.contains(10));
        assertTrue(tree.contains(30));
        assertTrue(tree.isBalanced(tree.getRoot()));
    }

    @Test
    void testBalanceAfterRemove() {
        // Create a tree that will become unbalanced after removal
        tree.insert(50);
        tree.insert(25);
        tree.insert(75);
        tree.insert(12);
        tree.insert(37);
        tree.insert(62);
        tree.insert(87);
        tree.insert(6);
        tree.insert(18);

        // Remove nodes to create potential imbalance
        tree.remove(87);
        tree.remove(75);

        // Verify tree is still balanced
        assertTrue(tree.isBalanced(tree.getRoot()));
    }

    @Test
    void testInsertAscendingOrder() {
        for (int i = 1; i <= 20; i++) {
            tree.insert(i);
            assertTrue(tree.isBalanced(tree.getRoot()), "Tree should be balanced after inserting " + i);
        }

        assertEquals(20, tree.size());
        assertEquals(1, tree.min());
        assertEquals(20, tree.max());
    }

    @Test
    void testInsertDescendingOrder() {
        for (int i = 20; i >= 1; i--) {
            tree.insert(i);
            assertTrue(tree.isBalanced(tree.getRoot()), "Tree should be balanced after inserting " + i);
        }

        assertEquals(20, tree.size());
        assertEquals(1, tree.min());
        assertEquals(20, tree.max());
    }

    @Test
    void testInsertRandomOrder() {
        int[] values = {42, 15, 88, 6, 27, 63, 94, 3, 9, 21, 36, 51, 79, 91, 99};

        for (int value : values) {
            tree.insert(value);
            assertTrue(tree.isBalanced(tree.getRoot()), "Tree should be balanced after inserting " + value);
        }

        assertEquals(values.length, tree.size());
        assertEquals(3, tree.min());
        assertEquals(99, tree.max());
    }

    @Test
    void testMinMaxCornerCases() {
        // Single node
        tree.insert(5);
        assertEquals(5, tree.min());
        assertEquals(5, tree.max());

        // Two nodes
        tree.insert(10);
        assertEquals(5, tree.min());
        assertEquals(10, tree.max());

        // Remove min
        tree.remove(5);
        assertEquals(10, tree.min());
        assertEquals(10, tree.max());
    }

    @Test
    void testComplexScenario() {
        // Insert a series of nodes
        tree.insert(50);
        tree.insert(25);
        tree.insert(75);
        tree.insert(12);
        tree.insert(37);
        tree.insert(62);
        tree.insert(87);

        // Verify tree state
        assertEquals(7, tree.size());
        assertEquals(12, tree.min());
        assertEquals(87, tree.max());
        assertTrue(tree.isBalanced(tree.getRoot()));

        // Remove nodes and verify balancing is maintained
        tree.remove(12);
        assertEquals(6, tree.size());
        assertTrue(tree.isBalanced(tree.getRoot()));

        tree.remove(87);
        assertEquals(5, tree.size());
        assertTrue(tree.isBalanced(tree.getRoot()));

        tree.remove(50);  // Remove root
        assertEquals(4, tree.size());
        assertTrue(tree.isBalanced(tree.getRoot()));

        // Insert more nodes
        tree.insert(40);
        tree.insert(35);

        assertEquals(6, tree.size());
        assertTrue(tree.isBalanced(tree.getRoot()));
    }

    @Test
    void testSearchCornerCases() {
        assertNull(tree.search(10));

        tree.insert(10);
        assertNotNull(tree.search(10));
        assertNull(tree.search(5));

        tree.insert(5);
        tree.insert(15);
        assertNotNull(tree.search(5));
        assertNotNull(tree.search(10));
        assertNotNull(tree.search(15));
        assertNull(tree.search(20));
    }

    @Test
    void testHeightCalculation() {
        tree.insert(10);
        assertEquals(1, tree.getRoot().getHeight());

        tree.insert(5);
        assertEquals(2, tree.getRoot().getHeight());

        tree.insert(15);
        assertEquals(2, tree.getRoot().getHeight());

        tree.insert(3);
        assertEquals(3, tree.getRoot().getHeight());
        assertTrue(tree.isBalanced(tree.getRoot()));
    }

    @Test
    void testRemoveAllNodes() {
        int[] values = {50, 25, 75, 12, 37, 62, 87};

        for (int value : values) {
            tree.insert(value);
        }

        assertEquals(values.length, tree.size());

        // Remove all nodes in order
        for (int value : values) {
            tree.remove(value);
            assertTrue(tree.isBalanced(tree.getRoot()));
        }

        assertTrue(tree.isEmpty());
        assertEquals(0, tree.size());
    }

    @Test
    void testStressTest() {
        // Insert a large number of elements to stress test the tree
        for (int i = 0; i < 1000; i++) {
            tree.insert(i);
            assertTrue(tree.isBalanced(tree.getRoot()));
        }

        assertEquals(1000, tree.size());

        // Remove elements randomly
        for (int i = 0; i < 1000; i += 2) {
            tree.remove(i);
            assertTrue(tree.isBalanced(tree.getRoot()));
        }

        assertEquals(500, tree.size());
    }
}

File: src/main/java/org/telran/lecture_13_avl/practice/AvlTreeTest.js
/**
 * Test suite for the AVL Tree implementation
 */
const { AvlTree } = require('./AvlTree');

/**
 * Runs a comprehensive test suite on the AVL Tree implementation
 */
function runTests() {
  let tree;

  /**
   * Initialize a fresh tree before each test
   */
  function beforeEach() {
    tree = new AvlTree();
  }

  /**
   * Assert that two values are equal
   * @param {*} actual - The actual value
   * @param {*} expected - The expected value
   * @param {string} message - Test description message
   */
  function assertEqual(actual, expected, message) {
    if (actual !== expected) {
      console.error(`❌ ${message} — Expected: ${expected}, Got: ${actual}`);
    } else {
      console.log(`✅ ${message}`);
    }
  }

  /**
   * Assert that a condition is true
   * @param {boolean} condition - The condition to test
   * @param {string} message - Test description message
   */
  function assertTrue(condition, message) {
    assertEqual(condition, true, message);
  }

  /**
   * Assert that a condition is false
   * @param {boolean} condition - The condition to test
   * @param {string} message - Test description message
   */
  function assertFalse(condition, message) {
    assertEqual(condition, false, message);
  }

  /**
   * Assert that a function throws an exception
   * @param {Function} fn - The function to test
   * @param {string} message - Test description message
   */
  function assertThrows(fn, message) {
    try {
      fn();
      console.error(`❌ ${message} — Expected exception, but none was thrown`);
    } catch {
      console.log(`✅ ${message}`);
    }
  }

  /**
   * Assert that a value is not null or undefined
   * @param {*} value - The value to test
   * @param {string} message - Test description message
   */
  function assertNotNull(value, message) {
    if (value === null || value === undefined) {
      console.error(`❌ ${message} — Value is null or undefined`);
    } else {
      console.log(`✅ ${message}`);
    }
  }

  // Tests for empty tree
  function testEmptyTree() {
    beforeEach();
    assertTrue(tree.isEmpty(), 'Tree should be empty');
    assertEqual(tree.size(), 0, 'Tree size should be 0');
    assertThrows(() => tree.min(), 'Calling min on empty tree should throw');
    assertThrows(() => tree.max(), 'Calling max on empty tree should throw');
    assertFalse(tree.contains(5), 'Tree should not contain 5');
    assertEqual(tree.search(5), null, 'Searching 5 in empty tree should return null');
  }

  // Test inserting a single node
  function testInsertSingleNode() {
    beforeEach();
    tree.insert(5);
    assertFalse(tree.isEmpty(), 'Tree should not be empty');
    assertEqual(tree.size(), 1, 'Tree size should be 1');
    assertEqual(tree.min(), 5, 'Min should be 5');
    assertEqual(tree.max(), 5, 'Max should be 5');
    assertTrue(tree.contains(5), 'Tree should contain 5');
    assertNotNull(tree.search(5), 'Searching 5 should return node');
    assertEqual(tree.search(5).value, 5, 'Value of found node should be 5');
    assertTrue(tree.isBalanced(), 'Tree should be balanced');
  }

  // Test inserting multiple nodes
  function testInsertMultipleNodes() {
    beforeEach();
    [5, 3, 7, 2, 4, 6, 8].forEach(v => tree.insert(v));
    assertEqual(tree.size(), 7, 'Tree size should be 7');
    assertEqual(tree.min(), 2, 'Min should be 2');
    assertEqual(tree.max(), 8, 'Max should be 8');
    assertTrue(tree.contains(3), 'Tree should contain 3');
    assertTrue(tree.contains(7), 'Tree should contain 7');
    assertFalse(tree.contains(1), 'Tree should not contain 1');
    assertFalse(tree.contains(9), 'Tree should not contain 9');
    assertTrue(tree.isBalanced(), 'Tree should be balanced');
  }

  // Test inserting duplicate values
  function testInsertDuplicates() {
    beforeEach();
    tree.insert(5);
    tree.insert(5);
    tree.insert(5);
    assertEqual(tree.size(), 1, 'Duplicate inserts should not increase size');
    assertEqual(tree.min(), 5, 'Min should be 5');
    assertEqual(tree.max(), 5, 'Max should be 5');
  }

  // Test left rotation (right-heavy tree)
  function testLeftRotation() {
    beforeEach();
    tree.insert(10);
    tree.insert(20);
    tree.insert(30);

    const root = tree.getRoot();
    assertEqual(root.value, 20, 'Root should be 20 after left rotation');
    assertEqual(root.left.value, 10, 'Left child should be 10');
    assertEqual(root.right.value, 30, 'Right child should be 30');
    assertTrue(tree.isBalanced(), 'Tree should be balanced');
  }

  // Test right rotation (left-heavy tree)
  function testRightRotation() {
    beforeEach();
    tree.insert(30);
    tree.insert(20);
    tree.insert(10);

    const root = tree.getRoot();
    assertEqual(root.value, 20, 'Root should be 20 after right rotation');
    assertEqual(root.left.value, 10, 'Left child should be 10');
    assertEqual(root.right.value, 30, 'Right child should be 30');
    assertTrue(tree.isBalanced(), 'Tree should be balanced');
  }

  // Test left-right rotation
  function testLeftRightRotation() {
    beforeEach();
    tree.insert(30);
    tree.insert(10);
    tree.insert(20);

    const root = tree.getRoot();
    assertEqual(root.value, 20, 'Root should be 20 after left-right rotation');
    assertEqual(root.left.value, 10, 'Left child should be 10');
    assertEqual(root.right.value, 30, 'Right child should be 30');
    assertTrue(tree.isBalanced(), 'Tree should be balanced');
  }

  // Test right-left rotation
  function testRightLeftRotation() {
    beforeEach();
    tree.insert(10);
    tree.insert(30);
    tree.insert(20);

    const root = tree.getRoot();
    assertEqual(root.value, 20, 'Root should be 20 after right-left rotation');
    assertEqual(root.left.value, 10, 'Left child should be 10');
    assertEqual(root.right.value, 30, 'Right child should be 30');
    assertTrue(tree.isBalanced(), 'Tree should be balanced');
  }

  // Test removing a single node
  function testRemoveSingleNode() {
    beforeEach();
    tree.insert(5);
    tree.remove(5);
    assertTrue(tree.isEmpty(), 'Tree should be empty after removal');
    assertEqual(tree.size(), 0, 'Tree size should be 0 after removal');
    assertFalse(tree.contains(5), 'Tree should not contain 5 after removal');
  }

  // Test removing a node that doesn't exist
  function testRemoveNonExistentNode() {
    beforeEach();
    tree.insert(5);
    tree.remove(10);
    assertEqual(tree.size(), 1, 'Size should not change after removing non-existent node');
    assertTrue(tree.contains(5), 'Tree should still contain 5');
  }

  // Test removing a leaf node
  function testRemoveLeafNode() {
    beforeEach();
    [20, 10, 30, 5].forEach(v => tree.insert(v));
    tree.remove(5);
    assertEqual(tree.size(), 3, 'Tree size should be 3 after removing leaf');
    assertFalse(tree.contains(5), 'Tree should not contain 5');
    assertTrue(tree.isBalanced(), 'Tree should be balanced');
  }

  // Test removing a node with one child
  function testRemoveNodeWithOneChild() {
    beforeEach();
    [20, 10, 30, 5].forEach(v => tree.insert(v));
    tree.remove(10);
    assertEqual(tree.size(), 3, 'Tree size should be 3 after removing node with one child');
    assertFalse(tree.contains(10), 'Tree should not contain 10');
    assertTrue(tree.contains(5), 'Tree should contain 5');
    assertTrue(tree.isBalanced(), 'Tree should be balanced');
  }

  // Test removing a node with two children
  function testRemoveNodeWithTwoChildren() {
    beforeEach();
    [20, 10, 30, 5, 15].forEach(v => tree.insert(v));
    tree.remove(10);
    assertEqual(tree.size(), 4, 'Tree size should be 4 after removing node with two children');
    assertFalse(tree.contains(10), 'Tree should not contain 10');
    console.log(JSON.stringify(tree));
    assertTrue(tree.contains(5), 'Tree should contain 5');
    assertTrue(tree.contains(15), 'Tree should contain 15');
    assertTrue(tree.isBalanced(), 'Tree should be balanced');
  }

  // Test complex imbalanced tree insertion
  function testComplexTreeInsertion() {
    beforeEach();
    // Insert values that require multiple rotations
    [50, 25, 75, 10, 30, 60, 80, 5, 15, 27, 35, 55, 65, 77, 85].forEach(v => tree.insert(v));
    
    assertTrue(tree.isBalanced(), 'Complex tree should be balanced after insertion');
    assertEqual(tree.size(), 15, 'Complex tree should have 15 nodes');
    assertEqual(tree.min(), 5, 'Min should be 5');
    assertEqual(tree.max(), 85, 'Max should be 85');
  }

  // Test complex tree removal
  function testComplexTreeRemoval() {
    beforeEach();
    // Insert values that require multiple rotations
    [50, 25, 75, 10, 30, 60, 80, 5, 15, 27, 35, 55, 65, 77, 85].forEach(v => tree.insert(v));
    
    // Remove some nodes to cause rebalancing
    tree.remove(25);
    tree.remove(75);
    tree.remove(50);
    
    assertTrue(tree.isBalanced(), 'Complex tree should remain balanced after removals');
    assertEqual(tree.size(), 12, 'Complex tree should have 12 nodes after removal');
    assertFalse(tree.contains(25), 'Tree should not contain 25');
    assertFalse(tree.contains(75), 'Tree should not contain 75');
    assertFalse(tree.contains(50), 'Tree should not contain 50');
  }

  // Run all tests
  const tests = [
    testEmptyTree,
    testInsertSingleNode,
    testInsertMultipleNodes,
    testInsertDuplicates,
    testLeftRotation,
    testRightRotation,
    testLeftRightRotation,
    testRightLeftRotation,
    testRemoveSingleNode,
    testRemoveNonExistentNode,
    testRemoveLeafNode,
    testRemoveNodeWithOneChild,
    testRemoveNodeWithTwoChildren,
    testComplexTreeInsertion,
    testComplexTreeRemoval
  ];

  console.log('\nRunning AVL Tree tests:\n');
  tests.forEach(testFn => {
    try {
      console.log(`Running test: ${testFn.name}`);
      testFn();
      console.log('');
    } catch (err) {
      console.error(`❌ Error during test ${testFn.name}: ${err.message}`);
      console.error(err.stack);
      console.log('');
    }
  });

  console.log('All tests completed.');
}

runTests();